### PR TITLE
Readme update: New sections AWS Cloud Discovery, Backpressure, Near Cache, Client System Properties

### PR DESCRIPTION
**Pull request by ihsandemir**
Added documentation for AWS Cloud Discovery, Backpressure, Near Cache and Client System Properties section.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -83,6 +84,7 @@
   * [7.8. Performance](#78-performance)
       * [7.8.1. Partition Aware](#781-partition-aware)
       * [7.8.2. Near Cache](#782-near-cache)
+        * [7.8.2.1 Configuring Near CacheConfiguring Near Cache](#7821-configuring-near-cache)
```
Comment (SaitTalhaNisanci) [inline README.md]: 7.8.2.1.

File: README.md [hunk 2]
```diff
@@ -450,6 +452,39 @@ You need to provide the IP address and port of at least one member in your clust
     hazelcast::client::ClientConfig config;
     config.getNetworkConfig().addAddress(hazelcast::client::Address("your server ip", 5701 /* your server port*/));
 ```
+### 1.4.3. Client System Properties
+
+While configuring your C++ client, you can use various system properties provided by Hazelcast to tune its clients. These properties can be set programmatically through `config.SetProperty` or by using an environment variable.
```
Comment (SaitTalhaNisanci) [inline README.md]: In go it is called `config` this should be `clientConfig` as it is called `clientConfig` in C++.

File: README.md [hunk 3]
```diff
@@ -1057,6 +1092,22 @@ for the client-cluster connection, you should set an SSL configuration. Please s
 
 As explained in the [TLS/SSL section](#61-tlsssl), Hazelcast members have key stores used to identify themselves (to other members) and Hazelcast C++ clients have certificate authorities used to define which members they can trust. 
 
+## 5.8. Enabling Hazelcast AWS Cloud Discovery
+
+The C++ client can discover the exiting Hazelcast servers in the Amazon AWS environment. The client queries the Amazon AWS environment using the [describe-instances] (http://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html) query of AWS. The client finds only the up and running instances and filters them based on the filter config provided at the ClientAwsConfig configuration.
```
Comment (SaitTalhaNisanci) [inline README.md]: If an instance is running, isnt it already up?  `The client finds only the up and running instances` -> `The client finds only the running instances`

File: README.md (hunk 2 above)
Comment (ihsandemir) [reply README.md]: This is the variable name (you can set anything) but not the class name, hence it should not matter.

File: README.md (hunk 3 above)
Comment (ihsandemir) [reply README.md]: The AWS instance may be up but there may be no Hazelcast server process running.

File: README.md [hunk 4]
```diff
@@ -1264,6 +1319,35 @@ You can set a timeout for retrying the operations sent to a member. This can be
...
+- Asynchronous backups: The asynchronous backups may be piling up.
+
+To prevent the system from crashing, Hazelcast provides back pressure. Back pressure works by:
+
+- limiting the number of concurrent operation invocations,
+- periodically making an async backup sync.
+
+Sometimes, e.g., when your servers are overloaded, you may want to slow down the client operations to the cluster. Then the client can be configured to wait until number of outstanding invocations whose reponses are not received to become less than a certain number. This is called "Client Backpressure". By default, the backpressure is disabled. There are a few properties which control the backpressure. These client configuration properties are:
+
+- `hazelcast.client.max.concurrent.invocations`: The maximum number of concurrent invocations allowed. To prevent the system from overloading, you can apply a constraint on the number of concurrent invocations. If the maximum number of concurrent invocations has been exceeded and a new invocation comes in, then Hazelcast will throw HazelcastOverloadException. By default this property is configured as INT32_MAX.
```
Comment (Serdaro) [inline README.md]: throw `HazelcastOverloadException`. By default, this property is configured as `INT32_MAX`.

File: README.md [hunk 5]
```diff
@@ -1264,6 +1319,35 @@ You can set a timeout for retrying the operations sent to a member. This can be
...
+- `hazelcast.client.invocation.backoff.timeout.millis`: Controls the maximum timeout in millis to wait for an invocation space to be available. If an invocation can't be made because there are too many pending invocations, then an exponential backoff is done to give the system time to deal with the backlog of invocations. This property controls how long an invocation is allowed to wait before getting `HazelcastOverloadException`. When set to -1 then `HazelcastOverloadException` is thrown immediately without any waiting. This is the default value.
```
Comment (Serdaro) [inline README.md]: ... maximum timeout in milliseconds to wait ...

File: README.md [hunk 6]
```diff
@@ -1264,6 +1319,35 @@ You can set a timeout for retrying the operations sent to a member. This can be
...
+
+For details of backpressure, see the [Backpressure section](https://docs.hazelcast.org/docs/latest/manual/html-single/index.html#back-pressure) in the Hazelcast IMDG Reference Manual.
```
Comment (Serdaro) [inline README.md]: For more information, see the [Back Pressure section](https://docs.hazelcast.org/docs/latest/manual/html-single/index.html#back-pressure) in the Hazelcast IMDG Reference Manual.

**Comment by ihsandemir** [comment]
verify-linux

File: README.md [hunk 7]
```diff
@@ -2333,6 +2417,91 @@ For more details, see the [PartitionAware section](https://docs.hazelcast.org/do
...
 
+Map  entries in Hazelcast are partitioned across the cluster members. Hazelcast clients do not have local data at all. Suppose you read the key `k` a number of times from a Hazelcast client or `k` is owned by another member in your cluster. Then each `map.get(k)` will be a remote operation, which creates a lot of network trips. If you have a data structure that is mostly read, then you should consider creating a local Near Cache, so that reads are sped up and less network traffic is created.
+
+These benefits do not come for free, please consider the following trade-offs:
+
+- Members with a Near Cache will have to hold the extra cached data, which increases memory consumption.
+- If invalidation is enabled and entries are updated frequently, then invalidations will be costly.
+- Near Cache breaks the strong consistency guarantees; you might be reading stale data.
+
+Near Cache is highly recommended for data structures that are mostly read.
```
Comment (Serdaro) [inline README.md]: Near Cache is highly recommended for maps that are mostly read.

File: README.md [hunk 8]
```diff
@@ -2333,6 +2417,91 @@ For more details, see the [PartitionAware section](https://docs.hazelcast.org/do
...
+
+In a client/server system you must enable the Near Cache separately on the client, without the need to configure it on the server. Please note that Near Cache configuration is specific to the server or client itself: a data structure on a server may not have Near Cache configured while the same data structure on a client may have Near Cache configured. They also can have different Near Cache configurations.
```
Comment (Serdaro) [inline README.md]: ... Please note that the Near Cache configuration ...

... a data structure on a server may not have a Near Cache configured while the same data structure on a client may have it configured. ...

File: README.md [hunk 9]
```diff
@@ -2333,6 +2417,91 @@ For more details, see the [PartitionAware section](https://docs.hazelcast.org/do
...
+
+If you are using Near Cache, you should take into account that your hits to the keys in the Near Cache are not reflected as hits to the original keys on the primary members. This has for example an impact on IMap's maximum idle seconds or time-to-live seconds expiration. Therefore, even though there is a hit on a key in Near Cache, your original key on the primary member may expire.
```
Comment (Serdaro) [inline README.md]: ... using the Near Cache, ...

... hit on a key in the Near Cache, ...

File: README.md [hunk 10]
```diff
@@ -2333,6 +2417,91 @@ For more details, see the [PartitionAware section](https://docs.hazelcast.org/do
...
+
+NOTE: Near Cache works only when you access data via `map.get(k)` method. Data returned using a predicate is not stored in the Near Cache.
```
Comment (Serdaro) [inline README.md]: ... data via the `map.get(k)` method.

File: README.md [hunk 11]
```diff
@@ -2333,6 +2417,91 @@ For more details, see the [PartitionAware section](https://docs.hazelcast.org/do
...
+
+A Near Cache can have its own `in-memory-format` which is independent of the `in-memory-format` of the data structure.
+
+#### 7.8.2.1 Configuring Near Cache
+
+Hazelcast Map can be configured to work with near cache enabled. You can enable the near cache for a map by adding a near cache configuration for that map. An example configuration for `myMap` is shown below.
```
Comment (Serdaro) [inline README.md]: You can enable the Near Cache on a Hazelcast Map by adding its configuration for that map. An example configuration for `myMap` is shown below.

File: README.md [hunk 12]
```diff
@@ -1057,6 +1096,22 @@ for the client-cluster connection, you should set an SSL configuration. Please s
...
+
+```C++
+clientConfig.getNetworkConfig().getAwsConfig().setEnabled(true).
+            setAccessKey(getenv("AWS_ACCESS_KEY_ID")).setSecretKey(getenv("AWS_SECRET_ACCESS_KEY")).
+            setTagKey("aws-test-tag").setTagValue("aws-tag-value-1").setSecurityGroupName("MySecureGroup").setRegion("us-east-1");
+```
+
+You need to enable the discovery by calling setEnabled(true). You can set your access key and secret in the config as shown in this example. You can filter the instances by setting which tags they have or by the security group setting. You can set the region for which the instances will be retrieved from, the default is to use us-east-1.
+ 
+The C++ client works the same way as the Java client. For details, see [AWSClient] (https://github.com/hazelcast/hazelcast-aws/blob/master/README.md) and [Hazelcast AWS Plugin] (https://github.com/hazelcast/hazelcast-aws/blob/master/README.md).
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply README.md]: Updated the first link as https://docs.hazelcast.org/docs/latest/manual/html-single/index.html#awsclient-configuration

File: README.md [hunk 13]
```diff
@@ -85,7 +85,7 @@
       * [7.8.1. Partition Aware](#781-partition-aware)
       * [7.8.2. Near Cache](#782-near-cache)
           * [7.8.2.1. Configuring Near Cache](#7821-configuring-near-cache)
-          * [7.8.2.2. Near Cache Example for Map](7822-near-cache-example-for-map)
+          * [7.8.2.2. Near Cache Example for Map](#7822-near-cache-example-for-map)
           * [7.8.2.3. Near Cache Eviction](7823-near-cache-eviction)
```
Comment (mdumandag) [inline README.md]: There are also missing '#'s in the other near cache items.

File: README.md [hunk 14]
```diff
@@ -2422,9 +2422,13 @@ Map  entries in Hazelcast are partitioned across the cluster members. Hazelcast
 These benefits do not come for free, please consider the following trade-offs:
 
 - Members with a Near Cache will have to hold the extra cached data, which increases memory consumption.
```
Comment (mdumandag) [inline README.md]: I thought that 'Clients with a Near Cache will have to hold the extra cached data, which increases memory consumption.'
could be a substitute for this item.

**Comment by ihsandemir** [comment]
verify-linux

**Pull request merged**